lib/matplotlib/backend_managers.py

```diff
@@ -14,6 +18,169 @@
 from matplotlib.rcsetup import validate_stringlist
 import matplotlib.backend_tools as tools
 
+from matplotlib import is_interactive
+from matplotlib import rcParams
+from matplotlib.figure import Figure
+from matplotlib.backends import get_backend
+
+
+class FigureManagerEvent(object):
+    """Event for when something happens to this figure manager.
+    i.e. the figure it controls gets closed
+
+    Attributes
+    ----------
+    signal : str
+        The name of the signal.
+
+    figure_manager : FigureManager
+        The figure manager that fired the event.
+    """
+    def __init__(self, signal, figure_manager):
+        self.name = signal
+        self.figure_manager = figure_manager
+
+
+class FigureManager(cbook.EventEmitter):
+    """
+    The FigureManager creates and wraps the necessary components to display a
+    figure, namely the Window, FigureCanvas and Toolbar.  It gets used whenever
+    you want the figure in a standalone window.
+
+    Parameters
+    ----------
+    figure : `matplotlib.figure.Figure`
+        The figure to manage.
+
+    num : int
+        The figure number.
+
+    Attributes
+    ----------
+
+    canvas : `matplotlib.backend_bases.FigureCanvasBase`
+        The GUI element on which we draw.
+
+    toolbar : `matplotlib.backend_bases.NavigationToolbar2`
+        The toolbar used for interacting with the figure.
+
+    window : `matplotlib.backend_bases.WindowBase`
+        The window that holds the canvas and toolbar.
+
+    num : int
+        The figure number.
+    """
+    def __init__(self, figure, num):
+        cbook.EventEmitter.__init__(self)
```

Comment:
I empathise a lot with what you say @efiring.  I definitely felt intimidated by the code base when I reported my first bug (#2616), I did a bit of initial digging and decided that that part of the codebase looked way too verbose, "spaghetti code", so yes, I just reported the bug and left.  Part of the reason for me wanting to work on these MEPs comes from the desire to make the codebase easier to understand, de-spaghettifi the code.

Decorators appear next on my list of things to learn, I also agree they look complicated.  Atm I treat them as voodoo, and as a sort of an embellishment to the code, but that has happened time and again with python (coming from the world of php/java/c) with things list comprehension etcetera, so I see decorators and super in amongst them.

The more I learn about ``super``, the more I like it, on the one hand I agree, it has a learning curve to it and that may put people off, on the other hand, if people just see it as voodoo and trust that it works (at least to begin with) and makes the code structure generally clearer then it may actually lower the boundaries to code contribution.

I found https://rhettinger.wordpress.com/2011/05/26/super-considered-super/ which helped to convert me, along with experimenting with some of the examples from there (don't worry most of the page gets filled up with comments, so much shorter then it looks).